---
title: "Breeding game: infer"
author: "Timoth√©e Flutre, Jacques David"
date: "`r format(Sys.time(), '%d/%m/%Y %H:%M:%S')`"
colorlinks: true
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: TRUE
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: TRUE
urlcolor: blue
---

<!--
This R chunk is used to set up some options.
-->
```{r setup, include=FALSE}
R.v.maj <- as.numeric(R.version$major)
R.v.min.1 <- as.numeric(strsplit(R.version$minor, "\\.")[[1]][1])
if(R.v.maj < 2 || (R.v.maj == 2 && R.v.min.1 < 15))
  stop("requires R >= 2.15", call.=FALSE)

suppressPackageStartupMessages(library(knitr))
opts_chunk$set(echo=TRUE, warning=TRUE, message=TRUE, cache=FALSE, fig.align="center")
```


# Overview

* https://github.com/timflutre/atelier-prediction-genomique

* Copyright 2016-2017 INRA, Montpellier SupAgro

* License: AGPL-3+

This R chunk is used to assess how much time it takes to execute the R code in this document until the end:
```{r time_0}
t0 <- proc.time()
```

Load the packages:
```{r load_pkg}
library(rutilstimflutre)
library(lme4)
library(sommer)
## library(QTLRel) # uncomment if the GEMMA software is not available
library(qqman) # for the "manhattan" function
library(rrBLUP)
```

Retrieve the directory paths:
```{r setup_dir}
root.dir <- "~/work2/atelier-prog-selection-2017"
setup <- getBreedingGameSetup(root.dir)
```


# Load all input files

Phenotypes:

* yield (trait1)

* quality (trait2)

* presence of disease symptoms (trait3)

```{r}
f <- paste0(setup$init.dir, "/phenos_coll.txt.gz")
phenos <- read.table(f, header=TRUE)
phenos$year <- as.factor(phenos$year)
phenos$plot <- as.factor(phenos$plot)
str(phenos)
```

Genotypes and SNP coordinates:
```{r}
f <- paste0(setup$init.dir, "/genos_coll-2014.txt.gz")
genos <- as.matrix(read.table(f, header=TRUE))
genos[1:3, 1:4]
snp.density <- "hd"
f <- paste0(setup$init.dir, "/snp_coords_", snp.density, ".txt.gz")
snp.coords <- read.table(f, header=TRUE, stringsAsFactors=FALSE)
nrow(snp.coords) == ncol(genos)
snp.coords <- snp.coords[colnames(genos),]
head(snp.coords)
```

Load truth:
```{r}
f <- paste0(setup$truth.dir, "/p0.RData")
load(f)
```


# Explore the data

Years with the presence of the pathogen on the site:
```{r}
(years.patho <- as.character(unique(phenos$year[phenos$pathogen])))
```

Summarize each trait:
```{r}
summary(phenos$trait1)
summary(phenos$trait2)
table(phenos$trait3[phenos$pathogen])
```

* large variation for trait1 and trait2

* several varieties have no symptoms the years with the pathogen, indicating that they are resistant

Plot the raw phenotypic data with no subsetting:
```{r}
hist(phenos$trait1, breaks="FD", main="trait 1", freq=FALSE)
hist(phenos$trait2, breaks="FD", main="trait 2", freq=FALSE)
```

Plot the raw phenotypic data by subsetting per year:
```{r}
boxplot(phenos$trait1 ~ phenos$year, las=1, main="trait 1", xlab="year",
        ylab="phenotypes")
abline(h=mean(phenos$trait1), lty=2)
boxplot(phenos$trait2 ~ phenos$year, las=1, main="trait 2", xlab="year",
        ylab="phenotypes")
abline(h=mean(phenos$trait2), lty=2)
```

* large `year` effect on the mean of each trait, but same variance

Plot the raw phenotypic data of trait1 and trait2 vs trait3 for the years at which the pathogen was observed:
```{r}
par(mfrow=c(1,2))
plot(jitter(phenos$trait3[phenos$pathogen]), phenos$trait1[phenos$pathogen],
     xlab="trait3", ylab="trait1", las=1)
plot(jitter(phenos$trait3[phenos$pathogen]), phenos$trait2[phenos$pathogen],
     xlab="trait3", ylab="trait2", las=1)
```

* right plot: whether or not individuals have symptoms (trait3), the mean product quality (trait2) doesn't change.

* left plot: the mean product yield (trait1) decreases strongly for individuals showing disease symptoms (i.e. the sensible varieties)

Plot the raw phenotypic data of trait1 vs trait2:
```{r}
plot(phenos$trait1, phenos$trait2, xlab="trait1", ylab="trait2",
     main="All years", las=1)
plot(phenos$trait1[! phenos$pathogen], phenos$trait2[! phenos$pathogen],
     xlab="trait1", ylab="trait2", main="Years without pathogen", las=1)
```

* there is a strong positive correlation between the raw phenotypic data of trait1 and trait2

Subset per year:
```{r}
for(year in levels(phenos$year))
  plot(phenos$trait1[phenos$year == year], phenos$trait2[phenos$year == year],
       xlab="trait1", ylab="trait2", main=year, las=1)
```


# Analysis of trait3

Extract the data for which genotypes and phenotypes are available:
```{r}
str(dat <- droplevels(phenos[phenos$year == 2014,]))
```

Find the true causal SNP of trait3 to highlight:
```{r}
true.qtl3 <- p0$trait3$qtn.id
```

Regress out known confounders:
```{r, eval=TRUE}
(fit3 <- lmer(formula=trait3 ~ 1 + (1|ind), data=dat))
blups.geno3 <- setNames(ranef(fit3)$ind[,"(Intercept)"],
                        rownames(ranef(fit3)$ind))
```

Launch GEMMA (if available):
```{r, eval=TRUE}
system.time(
    res3.gemma <- gemma(model="ulmm",
                        y=blups.geno3,
                        X=genos,
                        snp.coords=snp.coords,
                        ## chr.ids=c("chr1","chr2"),
                        W=matrix(rep(1, nrow(genos))),
                        out.dir=setup$truth.dir,
                        task.id="asso-coll-2014",
                        clean="all", verbose=1))
tmp <- data.frame(BP=snp.coords[rownames(res3.gemma$tests), "pos"],
                  CHR=as.numeric(gsub("chr", "",
                                      snp.coords[rownames(res3.gemma$tests), "chr"])),
                  P=res3.gemma$tests$p_wald,
                  SNP=rownames(res3.gemma$tests))
manhattan(x=tmp,
          chrlabs=paste0("chr", sort(unique(as.numeric(gsub("chr", "",
                                                            snp.coords$chr))))),
          main="GEMMA on BLUPs of trait3",
          suggestiveline=FALSE, genomewideline=FALSE, highlight=true.qtl3)
table(genos[, true.qtl3])
median(blups.geno3[genos[names(blups.geno3), true.qtl3] == 0])
median(blups.geno3[genos[names(blups.geno3), true.qtl3] == 2])
length(resist.varieties <- rownames(genos)[genos[,true.qtl3] == 2])

betterSummary(phenos$trait1[phenos$year == "2014"])
betterSummary(phenos$trait1[phenos$year == "2014" & ! phenos$ind %in% resist.varieties])
betterSummary(phenos$trait1[phenos$year == "2014" & phenos$ind %in% resist.varieties])
```

Launch QTLRel (if available):
```{r, eval=FALSE}
if(FALSE){
  afs <- colMeans(genos) / 2
  A.mark <- (genos - 1) %*% t(genos - 1) / (2 * sum(afs * (1 - afs)))
  Z <- model.matrix(~ dat$ind - 1)
  res3 <- list()
  res3$vc <- estVC(y=dat$trait3, x=Z,
                   v=list(AA=Z %*% A.mark %*% t(Z), DD=NULL, HH=NULL,
                          AD=NULL, MH=NULL, EE=diag(nrow(dat))))
  res3$scan <- scanOne(y=dat$trait3, x=Z, gdat=Z %*% genos,
                       vc=res3$vc, test="F", numGeno=TRUE)
  manhattan(data.frame(BP=snp.coords$pos,
                       CHR=as.numeric(gsub("chr", "", snp.coords$chr)),
                       P=res3$scan$p, SNP=rownames(snp.coords)),
            chrlabs=paste0("chr", sort(unique(as.numeric(gsub("chr", "",
                                                              snp.coords$chr))))),
            main="QTLRel", suggestiveline=FALSE, genomewideline=FALSE, ylim=c(0,200))
}
```


# Analysis of traits 1 and 2

## Separately

Note that the information in the `pathogen` column (in the `phenos` data.frame) is partially redundant with the `year` column, in interaction wih the `ind` column.
To ease the analysis, we add a new column indicating which variety seems resistant:
```{r}
phenos$resistant <- FALSE
phenos$resistant[phenos$pathogen & phenos$trait3 == 0] <- TRUE
```

### Trait1

Predict genotypic BLUPs for trait1:
```{r}
fit1.i <- lmer(trait1 ~ 1 + (1|ind), data=phenos)
extractAIC(fit1.i)
fit1.y.i <- lmer(trait1 ~ 1 + year + (1|ind), data=phenos)
extractAIC(fit1.y.i)
fit1.yi <- lmer(trait1 ~ 1 + year + (1|ind) + (1|year:ind), data=phenos)
extractAIC(fit1.yi)
fit1.y.r.i <- lmer(trait1 ~ 1 + year + resistant + (1|ind), data=phenos)
extractAIC(fit1.y.r.i)
fit1.y.p.i <- lmer(trait1 ~ 1 + year + plot + (1|ind), data=phenos)
extractAIC(fit1.y.p.i)
fit1.y.p.r.i <- lmer(trait1 ~ 1 + year + plot + resistant + (1|ind), data=phenos)
extractAIC(fit1.y.p.r.i)
fit1.y.pr.r.i <- lmer(trait1 ~ 1 + year + (1|plot) + resistant + (1|ind), data=phenos)
extractAIC(fit1.y.pr.r.i)
fit1 <- fit1.y.p.r.i
summary(fit1)
hist(residuals(fit1) / sigma(fit1)); abline(v=c(-2,0,2), lty=2)
plot(residuals(fit1) / sigma(fit1), fitted(fit1)); abline(v=c(-2,0,2), lty=2)
blups1 <- ranef(fit1.y.p.r.i, condVar=TRUE, drop=TRUE)$ind
```

### Trait2

Predict genotypic BLUPs for trait2:
```{r}
fit2.i <- lmer(trait2 ~ 1 + (1|ind), data=phenos)
extractAIC(fit2.i)
fit2.y.i <- lmer(trait2 ~ 1 + year + (1|ind), data=phenos)
extractAIC(fit2.y.i)
fit2.p.i <- lmer(trait2 ~ 1 + plot + (1|ind), data=phenos)
extractAIC(fit2.p.i)
fit2.y.p.i <- lmer(trait2 ~ 1 + year + plot + (1|ind), data=phenos)
extractAIC(fit2.y.p.i)
fit2.y.pr.i <- lmer(trait2 ~ 1 + year + (1|plot) + (1|ind), data=phenos)
extractAIC(fit2.y.pr.i)
fit2 <- fit2.y.i
summary(fit2)
hist(residuals(fit2) / sigma(fit2)); abline(v=c(-2,0,2), lty=2)
plot(residuals(fit2) / sigma(fit2), fitted(fit2)); abline(v=c(-2,0,2), lty=2)
blups2 <- ranef(fit2, condVar=TRUE, drop=TRUE)$ind
```

### Compare BLUPs for both traits

Compare BLUPs for both traits:
```{r}
all(names(blups1) == names(blups2))
cor(blups1, blups2)
regplot(blups1, blups2, main="Univariate LMM",
        xlab="BLUPs of trait1", ylab="BLUPs of trait2")
```

* there seems to be no (or a small negative) correlation between the genotypic values of both traits, but is it an artefact of the univariate inference?


## Jointly

Remove the pathogen effect only for trait1:
```{r}
fit1.r <- lm(trait1 ~ 1 + resistant, data=phenos)
phenos$res1 <- residuals(fit1.r)
```

Remove known confounders for both traits simultaneously:
```{r}
system.time(
    fit12 <- mmer(Y=phenos[, c("res1", "trait2")],
                  X=model.matrix(~ 1 + year, phenos),
                  Z=list(ind=list(Z=model.matrix(~ -1 + ind, phenos),
                                  K=diag(nlevels(phenos$ind)))),
                  method="NR", REML=TRUE, MVM=TRUE))
summary(fit12)
cor(fit12$u.hat$ind[,1], fit12$u.hat$ind[,2])
regplot(fit12$u.hat$ind[,1], fit12$u.hat$ind[,2], main="Multivariate LMM",
        xlab="BLUPs of trait1", ylab="BLUPs of trait2")
```

* there seems to be a strong negative correlation between the genotypic values of both traits

* at this stage, we still don't know the detailed architecture of the pleiotropy


# Appendix

```{r info}
t1 <- proc.time(); t1 - t0
print(sessionInfo(), locale=FALSE)
```
