---
title: "Breeding game: setup"
author: "Timoth√©e Flutre"
date: "`r format(Sys.time(), '%d/%m/%Y %H:%M:%S')`"
colorlinks: true
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: TRUE
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: TRUE
---

<!--
This R chunk is used to set up important options and load required packages.
-->
```{r setup, include=FALSE}
R.v.maj <- as.numeric(R.version$major)
R.v.min.1 <- as.numeric(strsplit(R.version$minor, "\\.")[[1]][1])
if(R.v.maj < 2 || (R.v.maj == 2 && R.v.min.1 < 15))
  stop("requires R >= 2.15", call.=FALSE)

suppressPackageStartupMessages(library(knitr))
opts_chunk$set(echo=TRUE, warning=TRUE, message=TRUE, cache=FALSE, fig.align="center")
```


# Overview

This document was conceived with the help of Jacques David (Montpellier SupAgro) and Philippe Brabant (AgroParisTech) for a master-level [workshop on genomic prediction and selection](https://github.com/timflutre/atelier-prediction-genomique).

It sets up a [serious game](https://github.com/timflutre/PlantSelBreedGame) to teach plant selective breeding.

* Copyright 2016-2018 INRA

* License: [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)


# Set up the environment

This R chunk is used to assess how much time it takes to execute the R code in this document until the end:
```{r time_0}
t0 <- proc.time()
```

Load the packages:
```{r load_pkg}
suppressPackageStartupMessages(library(DBI))
suppressPackageStartupMessages(library(RSQLite))
stopifnot(compareVersion("2.0",
                         as.character(packageVersion("RSQLite")))
          != 1)
suppressPackageStartupMessages(library(digest))
suppressPackageStartupMessages(library(MASS))
## library(LDheatmap) # optionnal
suppressPackageStartupMessages(library(rutilstimflutre))
stopifnot(compareVersion("0.156.6",
                         as.character(packageVersion("rutilstimflutre")))
          != 1)
```

Set up directories:
```{r setup_dir}
if(Sys.info()["user"] == "tflutre"){
  root.dir <- "~/work2/atelier-prog-selection-2018/data"
} else if(Sys.info()["user"] == "Julien DIOT"){
  root.dir <- "~/breeding-game/setup/data"
} else{
  root.dir <- "C:/<...>/data"
}
if(! dir.exists(root.dir))
  dir.create(root.dir)
setup <- setUpBreedingGame(root.dir=root.dir,
                           nb.breeders=2)
truth.dir <- setup$truth.dir
shared.dir <- setup$shared.dir
init.dir <- setup$init.dir
breeders <- setup$breeders
breeder.dirs <- setup$breeder.dirs
dbname <- setup$dbname
```


# Simulate haplotypes and genotypes

## Use the sequential coalescent with recombination

Nb of initial individuals:
```{r init_inds}
I <- 1000
ind.ids <- sprintf(fmt=paste0("ind%0", floor(log10(I))+1, "i"), 1:I)
```

Simulate haplotypes and genotypes of initial individuals:

* hyp: single pop

* hyp: no inter-chrom LD

* hyp: all chroms have the same length

* hyp: individuals are diploids

* scaling: measure time in units of 4 x N_e generations

```{r init_chrs}
nb.chrs <- 10
L <- 10^6     # chromosome length, in base pairs
mu <- 10^(-8) # neutral mutation rate in events / base / generation
(u <- mu * L) # neutral mutation rate in events / chrom / gen
c <- mu       # recomb rate in events / base / gen
(r <- c * L)  # recomb rate in events / chrom / gen
```

Simplest evolutionary scenario similar to real data from domesticated plants:
```{r evol_scenario}
##                             _____
##                            /
##       ____________________/      Nec
##                       _Neb_
##                       |    \_____
##            Nea        |
##                       |
##                       |
##       ________________|
##                       2   1      0
## time  <---------------+---+------+
Nea <- 5 * 10^4         # ancestral
Neb <- 10 * 10^2        # bottleneck
Nec <- Ne0 <- 5 * 10^3  # current
(theta <- 4 * Ne0 * u)  # scaled neutral mutation rate in events / chrom
(rho <- 4 * Ne0 * r)    # scaled recomb rate in events / chrom
T1 <- 8000
T2 <- T1 + 2000
(alpha <- - (4 * Ne0) / T1 * log(Neb / Ne0)) # exp growth rate from Nec to Neb
(cmd <- paste0("-eG ", T1/(4*Ne0), " ", alpha,
               " -eN ", T2/(4*Ne0), " ", Nea/Ne0))
```

Simulate genotypes via the coalescent:
```{r simul_geno_coalescent}
set.seed(65291)
f <- paste0(truth.dir, "/g0.RData")
if(! file.exists(f)){
  g0 <- simulCoalescent(nb.inds=I,
                        ind.ids=ind.ids,
                        nb.reps=nb.chrs,
                        pop.mut.rate=theta,
                        pop.recomb.rate=rho,
                        chrom.len=L,
                        other=cmd,
                        nb.pops=1,
                        get.trees=FALSE,
                        get.tmrca=FALSE,
                        permute.alleles=TRUE,
                        verbose=1)
  save(g0, file=f)
} else
  load(f)
tools::md5sum(path.expand(f))
file.info(f)$mtime
head(g0$snp.coords)
g0$haplos$chr1[1:4, 1:6]
g0$genos[1:2, 1:6]
```

## Look at (M)AFs, LD, A

Look at the AFs and MAFs:
```{r look_mafs}
afs <- estimSnpAf(X=g0$genos)
summary(afs)
plotHistAllelFreq(afs=afs,
                  main=paste0(length(afs), " SNPs"))
mafs <- estimSnpMaf(afs=afs)
summary(mafs)
sum(mafs >= 0.01) # nb of non-rare SNPs
plotHistMinAllelFreq(maf=mafs,
                     main=paste0(length(mafs), " SNPs"))
plotHistMinAllelFreq(maf=mafs[mafs >= 0.01],
                     main=paste0(sum(mafs >= 0.01), " SNPs (MAF >= 0.01)"))
```

Look at some haplotypes:
```{r look_haplos}
plotHaplosMatrix(g0$haplos$chr1[sort(sample.int(nrow(g0$haplos$chr1), 100)),
                                mafs[1:ncol(g0$haplos$chr1)] >= 0.2])
## calcAvgPwDiffBtwHaplos(g0$haplos$chr1) # slow
```

Estimate and plot pairwise LD per chromosome:
```{r estim_plot_ld}
chr <- "chr1"
min.maf <- 0.2
min.pos <- 0
max.pos <- L
(length(snps.tokeep <- rownames(g0$snp.coords[mafs >= min.maf &
                                              g0$snp.coords$chr == chr &
                                              g0$snp.coords$pos >= min.pos &
                                              g0$snp.coords$pos <= max.pos,])))
## snps.tokeep <- snps.tokeep[sort(sample(snps.tokeep, 100))]
(nrow(ld <- estimLd(X=g0$genos[,snps.tokeep],
                    snp.coords=g0$snp.coords[snps.tokeep,],
                    use.ldcorsv=FALSE)))
summary(ld$cor2)
snp.dist <- distSnpPairs(snp.pairs=ld[, c("loc1","loc2")],
                         snp.coords=g0$snp.coords[snps.tokeep,])
plotLd(snp.dist,
       ## ld$cor2, estim="r2",
       sqrt(ld$cor2), estim="r",
       main=paste0(length(snps.tokeep), " SNPs with MAF >= ", min.maf,
                   " on ", chr),
       use.density=TRUE,
       span=1/20,
       sample.size=2*I,
       Ne=Ne0, c=c,
       add.ohta.kimura=TRUE)
abline(v=500, lty=2)
## LDheatmap(gdat=cor(g0$genos[,snps.tokeep])^2,
##           genetic.distances=snp.dist, distances="physical",
##           title=paste0(length(snps.tokeep), " SNPs with MAF >= ", min.maf,
##                        " on ", chr),
##           add.map=FALSE, flip=FALSE)
```

Look at distances between consecutive SNPs:
```{r dist_btw_snps}
chr <- "chr3"
tmp <- distConsecutiveSnps(snp.coords=g0$snp.coords, only.chr=chr)
hist(tmp[[chr]], breaks="FD", xlab="in bp", las=1,
     main="Distances between consecutive SNPs")
abline(v=500, lty=2)
```

Thin SNPs:
```{r thin}
thin.snps <- thinSnps(method="coord", threshold=5*10^4,
                      snp.coords=g0$snp.coords[names(mafs)[mafs >= 0.1],])
length(thin.snps)
```

Estimate additive genomic relationships:
```{r estim_genrel}
A.vr <- estimGenRel(X=g0$genos[,thin.snps], method="vanraden1")
A <- A.vr
imageWithScale(z=A, main="Additive genomic relationships (A)")
inbred.coefs <- diag(A) - 1
summary(inbred.coefs) # compare to 0 if no inbreeding
hist(inbred.coefs, breaks="FD")
abline(v=0, lty=2, col="red")
coancestries <- A / 2
summary(diag(coancestries)) # compare to 0.5 for an outbred individual
hist(diag(coancestries), breaks="FD", xlim=c(0,1))
abline(v=0.5, lty=2, col="red")
summary(coancestries[upper.tri(coancestries)])
hist(coancestries[upper.tri(coancestries)], breaks="Sturges")
abline(v=0, lty=2, col="red")
```

## Structure into families

TODO: panmixia with 40 genotypes for 3 generations

## Choose SNPs for genotyping

Ascertain SNPs for the genotyping arrays (high and low density):

* hyp: mimick de novo sequencing of a few individuals

```{r ascertain_snps}
nb.inds.denovo <- 20
nb.snps.hd <- 10*10^3
nb.snps.ld <- 5*10^3
subset.inds <- sort(sample(rownames(g0$genos), nb.inds.denovo))
mafs.subset <- estimSnpMaf(g0$genos[subset.inds,])
summary(mafs.subset)
sum(mafs.subset >= 0.1)
subset.snps <- list()
subset.snps[["hd"]] <- sample(names(mafs.subset[mafs.subset >= 0.1]), nb.snps.hd)
subset.snps[["ld"]] <- sample(subset.snps[["hd"]], nb.snps.ld)
summary(mafs[subset.snps[["hd"]]])
summary(mafs[subset.snps[["ld"]]])
f <- paste0(init.dir, "/snp_coords_hd.txt.gz")
write.table(x=g0$snp.coords[subset.snps[["hd"]],],
            file=gzfile(f), quote=FALSE,
            sep="\t", row.names=TRUE, col.names=TRUE)
f <- paste0(init.dir, "/snp_coords_ld.txt.gz")
write.table(x=g0$snp.coords[subset.snps[["ld"]],],
            file=gzfile(f), quote=FALSE,
            sep="\t", row.names=TRUE, col.names=TRUE)
```

## Make lines

Make collection of lines via haplodiploidization:
```{r coll_lines}
f <- paste0(truth.dir, "/coll.Rdata")
coll <- list()
crosses <- data.frame(parent1=ind.ids,
                      parent2=NA,
                      child=sprintf(fmt=paste0("Coll%0", floor(log10(I))+1, "i"),
                                    1:I),
                      stringsAsFactors=FALSE)
loc.crossovers <- drawLocCrossovers(crosses, sapply(g0$haplos, ncol))
coll$haplos <- makeCrosses(g0$haplos, crosses, loc.crossovers)
coll$genos <- segSites2allDoses(seg.sites=coll$haplos, ind.ids=crosses$child,
                                snp.ids=colnames(g0$genos))
coll$snp.coords <- g0$snp.coords
save(coll, file=f)
tools::md5sum(path.expand(f))
file.info(f)$mtime
```

Free some memory:
```{r free_mem_genos}
## sort(sapply(ls(),function(x){object.size(get(x))}), decreasing=TRUE)
rm(g0, ld, snp.dist, mafs, loc.crossovers, mafs.subset, crosses,
   coancestries, A, A.vr)
gc()
```


# Simulate phenotypes

Simulate phenotypes of initial lines:

* hyp: only 3 traits are considered

* hyp: trait 1 has additive infinitesimal genetic architecture (e.g. yield)

* hyp: trait 2 has additive infinitesimal genetic architecture (e.g. quality)

* hyp: for traits 1 and 2, all SNPs (even the rarest) have a non-zero effect

* hyp: traits 1 and 2 are negatively genetically correlated (i.e. pleiotropy)

* hyp: "year" effects are different for traits 1 and 2, with the same variance but no covariance between traits or plots

* hyp: errors are different for traits 1 and 2, with the same variance but no covariance between traits, plots or years

* hyp: pathogens present every three years

* hyp: trait 3 has a single, major QTL (e.g. resistance)

* hyp: "trait3=0" means "absence of symptoms" and, very likely, presence of the resistance allele (but the years with pathogen, some genotypes without the resistant allele may exceptionnally show no symptom)

* hyp: the causal SNP is on the low-density chip

* hyp: the resistance-confering allele is the minor one, but isn't too rare

* hyp: the resistance-confering allele has a dominant effect

## Generic preparation

Make data.frame for initial phenotypes given to players:
```{r init_pheno}
nb.lines.per.year <- 150
nb.plots.per.line.per.year <- 2
nb.years <- 10
nb.plots <- nb.lines.per.year * nb.plots.per.line.per.year
line.ids <- getIndNamesFromHaplos(coll$haplos)[1:(150 + (10-1) * 75)]
first.year <- 2005
p0.df <- makeDfInitPhenos(nb.lines.per.year=nb.lines.per.year, nb.years=nb.years,
                          nb.plots.per.line.per.year=nb.plots.per.line.per.year,
                          first.year=first.year, line.ids)
head(p0.df)
str(p0.df)
summary(as.vector(table(p0.df$ind)))
```

Export genotypes for the players using high-density chip, but only for a subset of lines:
```{r export_geno}
f <- paste0(init.dir, "/genos_subset-coll.txt.gz")
subset.years <- as.numeric(levels(p0.df$year)[
  (floor(0.5*nlevels(p0.df$year))):nlevels(p0.df$year)])
subset.coll <- unique(as.character(p0.df$ind[p0.df$year %in% subset.years]))
write.table(x=coll$genos[subset.coll, subset.snps[["hd"]]],
            file=gzfile(f), quote=FALSE,
            sep="\t", row.names=TRUE, col.names=TRUE)
```

Thin to avoid LD between causal SNPs? No, keep all SNPs for the moment.
```{r choose_causal_snps}
## mafs <- estimSnpMaf(X=coll$genos[levels(p0.df$ind),])
## thin.snps <- thinSnps(method="coord", threshold=5*10^2,
##                       snp.coords=coll$snp.coords[names(mafs)[mafs >= 0.01],])
length(thin.snps <- colnames(coll$genos))
```

## Traits 1 and 2

Simulate the SNP effects for traits 1 and 2 jointly:
```{r simum_snp_effects_trait1_trait2}
set.seed(39452)
mu.traits12 <- c(trait1=40, trait2=14)
sigma.beta2 <- c(trait1=10^(-5), trait2=10^(-5))
(cv.traits12 <- sqrt(sigma.beta2 * (2 * sum(afs[thin.snps] * (1 - afs[thin.snps])))) / mu.traits12)
## mu.traits12 <- c(trait1=100, trait2=30)
## cv.traits12 <- c(trait1=0.05, trait2=0.02) # coefs of variation
## (exp.sigma.a <- mu.traits12 * cv.traits12)
## (sigma.beta2 <- exp.sigma.a^2 / (2 * sum(afs[thin.snps] * (1 - afs[thin.snps]))))
snp.effects12 <- simulSnpEffectsTraits12(snp.ids=thin.snps,
                                         sigma.beta2=sigma.beta2,
                                         prop.pleio=0.4, cor.pleio=-0.7)
```

Make phenotypes for traits 1 and 2 jointly:
```{r make_phenos_trait1_trait2}
h2.traits12 <- c(trait1=0.3, trait2=0.4)
sigma.alpha <- c(trait1=1.5, trait2=1)
sigma.alpha / mu.traits12
## h2.traits12 <- c(trait1=0.5, trait2=0.8)
## (sigma.alpha <- mu.traits12 * c(0.07, 0.1))
p0 <- simulTraits12(dat=p0.df, mu=mu.traits12,
                    sigma.alpha2=sigma.alpha^2,
                    X=coll$genos[levels(p0.df$ind), thin.snps],
                    Beta=snp.effects12$Beta,
                    h2=h2.traits12)
```

Do some visual checks:
```{r plots_phenos_trait1_trait2}
hist(p0$G.A[,1], breaks="FD", las=1, col="grey", border="white")
hist(p0$G.A[,2], breaks="FD", las=1, col="grey", border="white")
regplot(p0$G.A[,1], p0$G.A[,2], xlab="p0$G.A[,1]", ylab="p0$G.A[,2]", las=1)
hist(p0$Y[,1], breaks="FD", las=1, col="grey", border="white")
boxplot(p0$Y[,1] ~ p0.df$year, las=1, main="p0$Y[,1]")
abline(h=mean(p0$Y[,1]), lty=2)
hist(p0$Y[,2], breaks="FD", las=1, col="grey", border="white")
boxplot(p0$Y[,2] ~ p0.df$year, las=1, main="p0$Y[,2]")
abline(h=mean(p0$Y[,2]), lty=2)
regplot(p0$Y[,1], p0$Y[,2], xlab="p0$Y[,1]", ylab="p0$Y[,2]", las=1)
plot(p0$Y[,1], p0$Y[,2], xlab="p0$Y[,1]", ylab="p0$Y[,2]", las=1, type="n")
for(j in 1:nlevels(p0.df$year)){
  year <- levels(p0.df$year)[j]
  idx <- p0.df$year == year
  points(p0$Y[idx,1], p0$Y[idx,2], col=j)
  abline(lm(p0$Y[idx,2] ~ p0$Y[idx,1]), col=j)
}
```

## Trait 3

Make the "pathogen" column:
```{r patho_years}
p0.df$pathogen <- FALSE
first.year <- as.numeric(levels(p0.df$year)[1])
(years.with.patho <- seq(from=first.year, by=3, length.out=4))
p0.df$pathogen[p0.df$year %in% years.with.patho] <- TRUE
```

Make trait 3:
```{r make_trait3}
p0$trait3 <- simulTrait3(dat=p0.df, X=p0$X, afs=p0$afs,
                         subset.snps=subset.snps[["ld"]],
                         prob.resist.no.qtl=0.02)
table(p0$trait3$y)
```

## Finalize

Fill "phenotype" data.frame for players:
```{r fill_pheno_players}
p0.df$trait1.raw <- p0$Y[,1]
p0.df$trait2 <- p0$Y[,2]
p0.df$trait3 <- p0$trait3$y
```

Adjust trait 1 for yield loss in case of pathogen presence:
```{r adjust_trai1}
p0$prop.yield.loss <- 0.2
p0.df$trait1 <- p0.df$trait1.raw
idx <- p0.df$pathogen & (p0.df$trait3 == 1)
p0.df$trait1[idx] <- (1 - p0$prop.yield.loss) * p0.df$trait1[idx]
hist(p0.df$trait1, breaks="FD", las=1, col="grey", border="white")
boxplot(p0.df$trait1 ~ p0.df$year, las=1, main="p0.df$trait1")
abline(h=mean(p0.df$trait1), lty=2)
regplot(p0.df$trait1, p0.df$trait2, xlab="p0.df$trait1", ylab="p0.df$trait2",
        las=1)
```

Save data:
```{r save_phenos}
f <- paste0(truth.dir, "/p0.RData")
save(p0, file=f)
tools::md5sum(path.expand(f))
file.info(f)$mtime
```

Export "phenotype" data.frame for players:
```{r export_phenos}
f <- paste0(init.dir, "/phenos_coll.txt.gz")
write.table(x=p0.df[, -grep("raw", colnames(p0.df))],
            file=gzfile(f), quote=FALSE,
            sep="\t", row.names=FALSE, col.names=TRUE)
tools::md5sum(path.expand(f))
```


# Choose controls

At the end of the game, each breeder has to submit at least one new variety.
For them to be registered, they should be "better" than controls.
The controls are chosen so that they have:

* an "additive genotypic yield" among the 40% best lines,

* an "additive genotypic quality" above the mean of all lines,

* the resistant allele for some of them.

Retrieve the breeding values of each line in the collection:
```{r rank_genos}
BVs <- data.frame(trait1=p0$mu[1] + p0$G.A[,1],
                  trait2=p0$mu[2] + p0$G.A[,2],
                  trait3=! p0$X[, p0$trait3$qtn.id] %in% p0$trait3$resist.genos,
                  row.names=rownames(p0$X))
summary(BVs)
```

Identify potential controls:
```{r identify_potential_controls, eval=TRUE}
BVs <- BVs[order(BVs$trait1, decreasing=TRUE),]
length(idx1 <- 1:ceiling(0.4 * nrow(BVs)))
summary(BVs[idx1,])
length(idx2 <- which(BVs$trait2 >= mean(BVs$trait2)))
potential.controls <- BVs[intersect(idx1, idx2),]
summary(potential.controls)
```

Select five controls among them:
```{r choose_controls, eval=TRUE}
nb.controls <- 5
nb.ctls.resist <- 2
stopifnot(sum(potential.controls$trait3 == 0) >= nb.ctls.resist)
nb.ctls.sensitiv <- nb.controls - nb.ctls.resist
stopifnot(sum(potential.controls$trait3 != 0) >= nb.ctls.sensitiv)
length(pot.ctls.resist <- rownames(potential.controls)[potential.controls$trait3 == 0])
ctls.resist <- sample(pot.ctls.resist, nb.ctls.resist)
length(pot.ctls.sensitiv <- rownames(potential.controls)[potential.controls$trait3 != 0])
ctls.sensitiv <- sample(pot.ctls.sensitiv, nb.ctls.sensitiv)
controls <- c(ctls.resist, ctls.sensitiv)
BVs[controls,]
```

Plot how good they are compare to the rest of the collection:
```{r plot_controls_phenos}
plot(x=p0$mu[1] + p0$G.A[,1], y=p0$mu[2] + p0$G.A[,2], las=1, main="Controls")
abline(v=mean(p0$mu[1] + p0$G.A[,1]), h=mean(p0$mu[2] + p0$G.A[,2]), lty=2)
points(x=p0$mu[1] + p0$G.A[ctls.resist, 1],
       y=p0$mu[2] + p0$G.A[ctls.resist, 2],
       pch=19, col="green")
points(x=p0$mu[1] + p0$G.A[ctls.sensitiv, 1],
       y=p0$mu[2] + p0$G.A[ctls.sensitiv, 2],
       pch=17, col="green")
legend("topright", legend=c("resistant", "sensitive"), pch=c(19,17),
       col="green", bty="n")
plot(x=p0.df$trait1, y=p0.df$trait2, main="Controls")
points(x=p0.df$trait1[p0.df$ind %in% ctls.resist],
       y=p0.df$trait2[p0.df$ind %in% ctls.resist],
       pch=19, col="green")
points(x=p0.df$trait1[p0.df$ind %in% ctls.sensitiv],
       y=p0.df$trait2[p0.df$ind %in% ctls.sensitiv],
       pch=17, col="green")
legend("topleft", legend=c("resistant", "sensitive"), pch=c(19,17),
       col="green", bty="n")
```

Save controls for the players:
```{r savel_controls, eval=TRUE}
f <- paste0(init.dir, "/controls.txt")
write.table(x=controls, file=f, quote=FALSE, sep="\t", row.names=FALSE, col.names=FALSE)
```

Make and save phenotypes of controls in several additionnal plots the last few years to give a good reference to the players:
```{r phenos_controls, eval=TRUE}
nb.years.per.ctl <- 4
nb.plots.per.ctl <- 5
phectls.df <- NULL
latest.year <- as.numeric(levels(p0.df$year)[nlevels(p0.df$year)])
for(j in 1:nb.years.per.ctl){
  year <- latest.year - nb.years.per.ctl + j
  phectl.df <- makeDfPhenos(ind.ids=controls,
                            nb.plots.per.ind=rep(nb.plots.per.ctl,
                                                 length(controls)),
                            year=year,
                            pathogen=ifelse((year - 2005) %% 3 == 0,
                                            TRUE, FALSE))
  phectl.df$plot <- as.factor(as.numeric(phectl.df$plot) +
                              nlevels(p0.df$plot))
  phectl <- simulTraits12(dat=phectl.df, mu=p0$mu,
                          Alpha=p0$Alpha[as.character(year), , drop=FALSE],
                          X=p0$X[levels(phectl.df$ind),], Beta=p0$Beta,
                          sigma2=p0$sigma2, verbose=1)
  phectl$trait3 <- simulTrait3(dat=phectl.df, X=p0$X[levels(phectl.df$ind),],
                               qtn.id=p0$trait3$qtn.id,
                               resist.genos=p0$trait3$resist.genos,
                               prob.resist.no.qtl=p0$trait3$prob.resist.no.qtl,
                               verbose=0)
  phectl.df$trait1.raw <- phectl$Y[,1]
  phectl.df$trait2 <- phectl$Y[,2]
  phectl.df$trait3 <- phectl$trait3$y
  phectl.df$trait1 <- phectl.df$trait1.raw
  tmp <- (phectl.df$pathogen & as.logical(phectl.df$trait3))
  if(any(tmp))
    phectl.df$trait1[tmp] <- (1 - p0$prop.yield.loss) * phectl.df$trait1[tmp]
  phectl.df <- phectl.df[, -grep("raw", colnames(phectl.df))]
  print(summary(phectl.df))
  if(j == 1){
    phectls.df <- phectl.df
  } else
    phectls.df <- rbind(phectls.df, phectl.df)
}
regplot(phectls.df$trait1, phectls.df$trait2, xlab="phectls.df$trait1",
        ylab="phectls.df$trait2", las=1)
fout <- paste0(setup$init.dir, "/phenos_controls.txt.gz")
if(file.exists(fout))
  file.remove(fout)
write.table(x=phectls.df, file=gzfile(fout), quote=FALSE,
            sep="\t", row.names=FALSE, col.names=TRUE)
```

Free some memory:
```{r free_mem_phenos, eval=TRUE}
rm(p0)
gc()
```


# Set up database and example files

Create a database:
```{r setup_db}
db <- dbConnect(SQLite(), dbname=dbname)
dbListTables(db)
```

## Constants

Create the table with the constants (average phenotypes, unitary costs, etc):
```{r}
tbl <- "constants"
if(tbl %in% dbListTables(db)){
  query <- paste0("DROP TABLE ", tbl)
  res <- dbExecute(conn=db, query)
}
query <- paste0("CREATE TABLE ", tbl,
                " (item TEXT",
                ", value TEXT)")
res <- dbExecute(conn=db, query)
```

Fill the table with the constants used for the data simulation:
```{r}
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.chrs', '", nb.chrs, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('chr.length', '", L, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.snps', '", length(thin.snps), "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('mu.trait1', '", mu.traits12[1], "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('mu.trait2', '", mu.traits12[2], "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.inds.denovo', '", nb.inds.denovo, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.snps.hd', '", nb.snps.hd, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.snps.ld', '", nb.snps.ld, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.phenotyped.coll', '", nlevels(p0.df$ind), "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.genotyped.coll', '", length(subset.coll), "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('max.nb.haplodiplos', '", 700, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.plots', '", nb.plots, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.controls', '", nb.controls, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.years.per.ctl', '", nb.years.per.ctl, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('nb.plots.per.ctl', '", nb.plots.per.ctl, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('register.min.trait1', '", 1.03, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('register.min.trait2', '", 14, "')")
res <- dbExecute(conn=db, query)
```

Fill the table with the constants specifying the time delays (in months):
```{r}
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('first.year', '", first.year, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('max.upload.pheno.field', '", "05-31", "')") # May 31 of each year
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.pheno.field', '", 4, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.pheno.patho', '", 3, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.allof', '", 4, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.autof', '", 4, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.haplodiplo', '", 12, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.geno.hd', '", 1, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.geno.ld', '", 1, "')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('duration.geno.single', '", 1, "')")
res <- dbExecute(conn=db, query)
```

Fill the table with the constants specifying the costs:
```{r}
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.pheno.field', '", 50, "')") # in Mendels
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.pheno.patho', '", 1/10, "')") # compare to one plot
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.allof', '", 1/10, "')") # idem
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.autof', '", 1/25, "')") # idem
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.haplodiplo', '", 1, "')") # idem
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.geno.hd', '", 1, "')") # idem
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.geno.ld', '", 1/3, "')") # idem
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.geno.single', '", 1/50, "')") # idem
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('cost.register', '", 4, "')") # idem
res <- dbExecute(conn=db, query)
```

## Log

Create the table for logging:
```{r create_table_log}
tbl <- "log"
if(tbl %in% dbListTables(db)){
  query <- paste0("DROP TABLE ", tbl)
  res <- dbExecute(conn=db, query)
}
query <- paste0("CREATE TABLE ", tbl,
                " (breeder TEXT",
                ", request_date TEXT", # ISO-8601
                ", task TEXT",
                ", quantity INT",
                ", time TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL)")
res <- dbExecute(conn=db, query)
```

## Sessions

Create the table describing the game sessions:
```{r}
tbl <- "sessions"
if(tbl %in% dbListTables(db)){
  query <- paste0("DROP TABLE ", tbl)
  res <- dbExecute(conn=db, query)
}
query <- paste0("CREATE TABLE ", tbl,
                " (num INT",
                ", start TEXT",
                ", end TEXT",
                ", year_time INT)") # in minutes
res <- dbExecute(conn=db, query)
```

Fill the table:
```{r}
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('1'",
                ", '2018-02-15 09:00:00'",
                ", '2018-02-15 12:00:00'",
                ", '60')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('2'",
                ", '2018-02-15 14:00:00'",
                ", '2018-02-15 17:00:00'",
                ", '60')")
res <- dbExecute(conn=db, query)
query <- paste0("INSERT INTO ", tbl, " VALUES",
                " ('3'",
                ", '2018-02-16 09:00:00'",
                ", '2018-02-16 11:40:00'",
                ", '40')")
res <- dbExecute(conn=db, query)
```

## Breeders

Create the table for breeder information, and insert the breeder(s):
```{r create_table_breeders}
tbl <- "breeders"
if(tbl %in% dbListTables(db)){
  query <- paste0("DROP TABLE ", tbl)
  res <- dbExecute(conn=db, query)
}
query <- paste0("CREATE TABLE ", tbl,
                " (name TEXT",
                ", status TEXT",
                ", h_psw TEXT)")
res <- dbExecute(conn=db, query)
hashed.psw <- digest("", "md5", serialize=FALSE)
for(breeder in breeders){
  query <- paste0("INSERT INTO ", tbl, " VALUES",
                  " ('", breeder)
  if(breeder == "test"){
    query <- paste0(query,
                    "', 'game master",
                    "', '", hashed.psw, "')")
  } else
    query <- paste0(query,
                    "', 'player",
                    "', '", hashed.psw, "')")
  res <- dbExecute(conn=db, query)
}
```

## Plant material

Create one table per breeder for their plant material:
```{r create_tables_plant_material}
for(breeder in breeders){
  tbl <- paste0("plant_material_", breeder)
  if(tbl %in% dbListTables(db)){
    query <- paste0("DROP TABLE ", tbl)
    res <- dbExecute(conn=db, query)
  }
  query <- paste0("CREATE TABLE ", tbl,
                  " (parent1 TEXT",
                  ", parent2 TEXT",
                  ", child TEXT PRIMARY KEY",
                  ", avail_from TEXT)")
  res <- dbExecute(conn=db, query)
}
dbListTables(db)
```

For each individual initially given to the players, save it into the table, and save its haplotypes into a file (slow):
```{r fill_table}
f <- paste0(truth.dir, "/coll.Rdata")
load(f)
ind.ids <- rownames(coll$genos)
tmp.ids <- gsub("Coll", "Ind", ind.ids) # dummy identifiers for the parent1's
for(i in 1:length(ind.ids)){
  ind.id <- ind.ids[i]
  ind <- list(haplos=getHaplosInd(haplos=coll$haplos, ind.name=ind.id))
  f <- paste0(truth.dir, "/", ind.id, "_haplos.RData")
  save(ind, file=f)
  for(breeder in breeders){
    fto <- paste0(truth.dir, "/", breeder, "/", ind.id, "_haplos.RData")
    if(! file.exists(fto))
      file.symlink(f, fto)
    tbl <- paste0("plant_material_", breeder)
    query <- paste0("INSERT INTO ", tbl, " VALUES",
                    " ('", tmp.ids[i], "'",
                    ", '", NA, "'",
                    ", '", ind.id, "'",
                    ", '", first.year, "-01-01 00:00:00')") # ISO-8601
    res <- dbExecute(conn=db, query)
  }
}
```

## Example files

Make an example of a "plant material" file:
```{r ex_cross}
(plant.examples <- makeExamplePlantFile(init.dir))
```

Make an example of an "data" file:
```{r ex_data}
(data.examples <- makeExampleDataFile(init.dir))
```


# Export ZIP archive

Make a zip archive of the whole directory:
```{r make_zip}
cwd <- getwd()
setwd(dirname(setup$root.dir))
system.time(
    zip(zipfile="data.zip",
        files=list.files(path="data", full.names=TRUE, recursive=TRUE,
                         include.dirs=TRUE)))
setwd(cwd)
```


# Appendix

```{r info}
t1 <- proc.time(); t1 - t0
print(sessionInfo(), locale=FALSE)
```
