---
title: "Breeding game: infer"
author: "Timoth√©e Flutre, Philippe Brabant, Jacques David"
date: "`r format(Sys.time(), '%d/%m/%Y %H:%M:%S')`"
colorlinks: true
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    number_sections: TRUE
  pdf_document:
    toc: true
    toc_depth: 4
    number_sections: TRUE
---

<!--
This R chunk is used to set up some options.
-->
```{r setup, include=FALSE}
R.v.maj <- as.numeric(R.version$major)
R.v.min.1 <- as.numeric(strsplit(R.version$minor, "\\.")[[1]][1])
if(R.v.maj < 2 || (R.v.maj == 2 && R.v.min.1 < 15))
  stop("requires R >= 2.15", call.=FALSE)

suppressPackageStartupMessages(library(knitr))
opts_chunk$set(echo=TRUE, warning=TRUE, message=TRUE, cache=FALSE, fig.align="center")
```


# Overview

* https://github.com/timflutre/atelier-prediction-genomique

* Copyright 2016-2018 INRA, Montpellier SupAgro, AgroParisTech

* License: [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)


# Set up the environment

This R chunk is used to assess how much time it takes to execute the R code in this document until the end:
```{r time_0}
t0 <- proc.time()
```

Load the packages:
```{r load_pkg}
suppressPackageStartupMessages(library(lattice))
suppressPackageStartupMessages(library(QTLRel))
suppressPackageStartupMessages(library(qqman))
suppressPackageStartupMessages(library(lme4))
## suppressPackageStartupMessages(library(sommer))
## suppressPackageStartupMessages(library(breedR))
suppressPackageStartupMessages(library(rrBLUP))
stopifnot(compareVersion("4.5",
                         as.character(packageVersion("rrBLUP")))
          != 1)
suppressPackageStartupMessages(library(cvTools))
suppressPackageStartupMessages(library(rutilstimflutre))
stopifnot(compareVersion("0.156.4",
                         as.character(packageVersion("rutilstimflutre")))
          != 1)
```

Retrieve the directory paths:
```{r setup_dir}
if(Sys.info()["user"] == "tflutre"){
  root.dir <- "~/work2/atelier-prog-selection-2018/data"
} else{
  stop("you need to add your path to this Rmd!")
}
setup <- getBreedingGameSetup(root.dir)
```


# Load all input files

## Phenotypes

* `trait1`: flower in kg/ha (quantity)

* `trait2`: sepmetin in g/kg of flower (quality)

* `trait3=1`: presence of disease symptoms

Collection:
```{r}
f <- paste0(setup$init.dir, "/phenos_coll.txt.gz")
tools::md5sum(path.expand(f))
phenos <- read.table(f, header=TRUE)
phenos$year <- as.factor(phenos$year)
phenos$plot <- as.factor(phenos$plot)
str(phenos)
```

Controls:
```{r}
f <- paste0(setup$init.dir, "/phenos_controls.txt.gz")
tools::md5sum(path.expand(f))
phectls <- read.table(f, header=TRUE)
phectls$year <- as.factor(phectls$year)
phectls$plot <- as.factor(phectls$plot)
str(phectls)
```

## Genotypes and SNP coordinates

```{r}
f <- paste0(setup$init.dir, "/genos_subset-coll.txt.gz")
tools::md5sum(path.expand(f))
genos <- as.matrix(read.table(f, header=TRUE))
dim(genos)
genos[1:3, 1:4]
snp.density <- "hd"
f <- paste0(setup$init.dir, "/snp_coords_", snp.density, ".txt.gz")
snp.coords <- read.table(f, header=TRUE, stringsAsFactors=FALSE)
stopifnot(nrow(snp.coords) == ncol(genos))
snp.coords <- snp.coords[colnames(genos),]
dim(snp.coords)
head(snp.coords)
```

## Truth

```{r}
f <- paste0(setup$truth.dir, "/p0.RData")
tools::md5sum(path.expand(f))
load(f)
```


# Explore the data

## Numerically

Years with the presence of the pathogen on the site:
```{r}
(years.patho <- as.character(unique(phenos$year[phenos$pathogen])))
```

Summarize each trait:
```{r}
summary(phenos$trait1)
summary(phenos$trait2)
summary(phenos$trait1 * phenos$trait2 / 1000)
table(phenos$trait3[phenos$pathogen])
```

* large variation for trait1 and trait2

* several varieties have no symptoms (trait3=0) the years with the pathogen, indicating that they are resistant

Effects of trait3 on the others:
```{r}
summary(phenos$trait1[! phenos$pathogen])
summary(phenos$trait1[phenos$pathogen])
summary(phenos$trait2[! phenos$pathogen])
summary(phenos$trait2[phenos$pathogen])
```

* the pathogen seems to have an effect on both traits

Correlations between traits:
```{r}
cor(phenos$trait1, phenos$trait2, method="pearson")
cor(phenos$trait1, phenos$trait2, method="spearman")
cor(phenos$trait1[! phenos$pathogen], phenos$trait2[! phenos$pathogen],
    method="pearson")
cor(phenos$trait1[phenos$pathogen], phenos$trait2[phenos$pathogen],
    method="spearman")
```

* hard to interpret

Maybe there is a strong year effect:
```{r}
do.call(rbind, tapply(phenos$trait1, list(phenos$year), summary))
do.call(rbind, tapply(phenos$trait1[! phenos$pathogen],
                      list(droplevels(phenos$year[! phenos$pathogen])),
                      summary))
do.call(rbind, tapply(phenos$trait2, list(phenos$year), summary))
do.call(rbind, tapply(phenos$trait2[! phenos$pathogen],
                      list(droplevels(phenos$year[! phenos$pathogen])),
                      summary))
```

* indeed the year effect is quite strong

## Graphically

Plot the raw phenotypic data with no subsetting:
```{r}
hist(phenos$trait1, breaks="FD", main="trait 1", las=1)
hist(phenos$trait2, breaks="FD", main="trait 2", las=1)
```

Plot the raw phenotypic data, subsetting per year:
```{r}
boxplot(phenos$trait1 ~ phenos$year, las=1, main="trait 1", xlab="year",
        ylab="phenotypes")
abline(h=mean(phenos$trait1), lty=2)
boxplot(phenos$trait2 ~ phenos$year, las=1, main="trait 2", xlab="year",
        ylab="phenotypes")
abline(h=mean(phenos$trait2), lty=2)
```

* clear `year` effect on the mean of each trait, but same variance

Plot the raw phenotypic data of trait1 and trait2 vs trait3 for the years at which the pathogen was observed:
```{r, fig.width=14, fig.height=9}
par(mfrow=c(1,2))
plot(jitter(phenos$trait3[phenos$pathogen]), phenos$trait1[phenos$pathogen],
     xlab="trait3", ylab="trait1", las=1)
plot(jitter(phenos$trait3[phenos$pathogen]), phenos$trait2[phenos$pathogen],
     xlab="trait3", ylab="trait2", las=1)
```

* left plot: trait1 decreases strongly for individuals showing disease symptoms (`trait3=1`)

* right plot: trait2 doesn't change whether individuals have symptoms or not

Plot the raw phenotypic data of trait1 vs trait2:
```{r}
plot(phenos$trait1, phenos$trait2, xlab="trait1", ylab="trait2",
     main="All years", las=1)
abline(lm(trait2 ~ trait1, data=phenos), col="red")
plot(phenos$trait1[! phenos$pathogen], phenos$trait2[! phenos$pathogen],
     xlab="trait1", ylab="trait2", main="Years without pathogen", las=1)
abline(lm(trait2 ~ trait1, data=phenos[! phenos$pathogen,]), col="red")
```

Same but distinguish each year by different colors:
```{r}
plot(phenos$trait1, phenos$trait2, xlab="trait1", ylab="trait2",
     main="All years", las=1, type="n")
for(j in 1:nlevels(phenos$year)){
  year <- levels(phenos$year)[j]
  idx <- which(phenos$year == year)
  points(phenos$trait1[idx], phenos$trait2[idx], col=j)
  abline(lm(trait2 ~ trait1, data=phenos[idx,]), col=j)
}
```

Same but only the years without pathogen:
```{r}
plot(phenos$trait1, phenos$trait2, xlab="trait1", ylab="trait2",
     main="Years without pathogen", las=1, type="n")
for(j in 1:nlevels(phenos$year)){
  year <- levels(phenos$year)[j]
  if(year %in% years.patho)
    next
  idx <- which(phenos$year == year)
  points(phenos$trait1[idx], phenos$trait2[idx], col=j)
  abline(lm(trait2 ~ trait1, data=phenos[idx,]), col=j)
}
```

* negative correlation between trait1 and trait2, especially visible the years without pathogen

## Raw data subsetted per year

```{r, eval=TRUE}
for(year in levels(phenos$year)){
  x <- phenos$trait1[phenos$year == year]
  y <- phenos$trait2[phenos$year == year]
  ## plot(x, y, xlab="trait1", ylab="trait2", las=1,
  ##      main=paste0(year, ifelse(year %in% years.patho, " (pathogen)", "")))
  ## abline(lm(y ~ x), col="red")
  ## legend("bottomright", legend=paste0("cor=", format(cor(x, y), digits=2)),
  ##        col="red", lty=1, bty="n")
  regplot(x, y, xlab="trait1", ylab="trait2",
          main=paste0(year, ifelse(year %in% years.patho, " (pathogen)", "")))
}
```

## Averaged data subsetted per year

```{r, eval=TRUE}
for(year in levels(phenos$year)){
  x <- droplevels(phenos[phenos$year == year, c("ind","trait1")])
  x <- tapply(x$trait1, list(x$ind), mean)
  x <- setNames(x, names(x))
  y <- droplevels(phenos[phenos$year == year, c("ind","trait2")])
  y <- tapply(y$trait2, list(y$ind), mean)
  y <- setNames(y, names(y))
  ## plot(x, y, xlab="trait1", ylab="trait2", las=1,
  ##      main=paste0(year, ifelse(year %in% years.patho, " (pathogen)", "")))
  ## abline(lm(y ~ x), col="red")
  ## legend("bottomright", legend=paste0("cor=", format(cor(x, y), digits=2)),
  ##        col="red", lty=1, bty="n")
  regplot(x, y, xlab="trait1 averaged per ind",
          ylab="trait2 averaged per ind",
          main=paste0(year, ifelse(year %in% years.patho, " (pathogen)", "")))
}
```


# Analysis of trait3

Extract the data of the latest year for which genotypes and phenotypes are available:
```{r}
last.year <- "2014"
dat <- droplevels(phenos[phenos$year == last.year,])
str(dat)
```

## Linear model

### Model fitting

```{r}
fit.i <- lm(formula=trait3 ~ 1 + ind, data=dat)
```

TODO

## Linear mixed model

### Model fitting

```{r}
fit.i <- lmer(formula=trait3 ~ 1 + (1|ind), data=dat, REML=FALSE)
```

TODO: check if presence of spatial heterogeneity?

### Model comparison

```{r}
AIC(fit.i)
BIC(fit.i)
```

### Model selection

```{r}
bestmod <- fit.i
```

### Diagnostics

Prepare to diagnose the best model:
```{r}
fit.all <- cbind(dat,
                 response=dat$trait3,
                 cond.res=residuals(bestmod),
                 scl.cond.res=residuals(bestmod) / sigma(bestmod),
                 fitted=fitted(bestmod))
ind.cond.mode <- ranef(bestmod, condVar=TRUE, drop=TRUE)$ind
var.ind.cond.mode <- setNames(attr(ind.cond.mode, "postVar"),
                              names(ind.cond.mode))
```

Check the homoscedasticity of the errors:
```{r}
low.bound.norm.95 <- qnorm(p=0.025, mean=0, sd=1, lower.tail=TRUE)
x.lim <- max(abs(fit.all$scl.cond.res), na.rm=TRUE)
plot(x=jitter(fit.all$scl.cond.res), y=jitter(fit.all$fitted), las=1,
     xlim=c(-x.lim, x.lim), main=paste0("trait3 in ", last.year))
abline(v=0, lty=2)
abline(v=c(low.bound.norm.95, abs(low.bound.norm.95)), lty=3)
```

Check the normality of the errors:
```{r}
shapiro.test(fit.all$scl.cond.res)
qqnorm(y=fit.all$scl.cond.res,
       main=paste0("trait3 in ", last.year, ": Normal Q-Q plot of scl.cond.res"))
qqline(y=fit.all$scl.cond.res, col="red")
```

Check the absence of outlying individuals:
```{r}
x.lim <- max(abs(ind.cond.mode))
par(mar=c(5,6,4,1))
plot(x=ind.cond.mode, y=1:length(ind.cond.mode),
     xlim=c(-x.lim, x.lim), main=paste0("trait3 in ", last.year),
     yaxt="n", ylab="")
axis(side=2, at=1:length(ind.cond.mode), labels=names(ind.cond.mode), las=1)
abline(v=0, lty=2)
idx <- which.max(ind.cond.mode)
text(x=ind.cond.mode[idx], y=idx, labels=names(idx), pos=2)
idx <- which.min(ind.cond.mode)
text(x=ind.cond.mode[idx], y=idx, labels=names(idx), pos=4)
```

Check the normality of the random variables:
```{r}
qqnorm(y=ind.cond.mode, main="Normal Q-Q plot of ind.cond.mode", asp=1)
qqline(y=ind.cond.mode, col="red")
```

Check the independence between errors and random variables:
```{r}
x.lim <- max(abs(fit.all$scl.cond.res))
dotplot(ind ~ scl.cond.res, data=fit.all,
        xlim=c(-x.lim, x.lim), main=paste0("trait3 in ", last.year),
        panel=function(x,y,...){
          panel.abline(v=0, lty=2)
          panel.abline(v=c(low.bound.norm.95, abs(low.bound.norm.95)), lty=3)
          panel.dotplot(x,y,...)
        })
```

### Inference

Estimate variance components by REML:
```{r}
bestmod <- lmer(formula=formula(bestmod),
                data=dat, na.action=na.exclude, REML=TRUE)
summary(bestmod)
```

Compute the confidence interval for variance components:
```{r}
prof <- stats::profile(fitted=bestmod, signames=FALSE)
(ci <- stats::confint(object=prof, level=0.95))
```

Retrieve the true causal SNP of trait3 to highlight it in the Manhattan plot:
```{r}
(true.qtl3 <- p0$trait3$qtn.id)
```

### QTL detection with GEMMA

Launch [GEMMA](https://github.com/xiangzhou/GEMMA) (better and quicker than QTLRel, but only on Linux), if available:
```{r, eval=TRUE}
if(file.exists(Sys.which("gemma"))){
  print(system.time(
      res3.gemma <- gemma(model="ulmm",
                          y=ind.cond.mode,
                          X=genos[names(ind.cond.mode),],
                          snp.coords=snp.coords,
                          ## chr.ids=snp.coords[true.qtl3, "chr"],
                          W=matrix(rep(1, length(ind.cond.mode))),
                          out.dir=setup$truth.dir,
                          task.id=paste0("asso-coll-", last.year),
                          clean="all", verbose=1)))
  tmp <- data.frame(BP=snp.coords[rownames(res3.gemma$tests), "pos"],
                    CHR=as.numeric(gsub("chr", "",
                                        snp.coords[rownames(res3.gemma$tests),
                                                   "chr"])),
                    P=res3.gemma$tests$p_wald,
                    SNP=rownames(res3.gemma$tests))
  print(best.snp <- as.character(tmp$SNP[which.min(tmp$P)]))
  print(tmp[tmp$SNP %in% c(true.qtl3, best.snp),])
  qqman::manhattan(x=tmp,
                   chrlabs=paste0("chr",
                                  sort(unique(as.numeric(gsub("chr", "",
                                                              snp.coords$chr))))),
                   main="GEMMA on BLUPs of trait3",
                   suggestiveline=FALSE, genomewideline=FALSE, highlight=true.qtl3)
  print(table(genos[names(ind.cond.mode), true.qtl3]))
  print(median(ind.cond.mode[genos[names(ind.cond.mode), true.qtl3] == 0]))
  print(median(ind.cond.mode[genos[names(ind.cond.mode), true.qtl3] == 2]))
  print(length(resist.varieties <- rownames(genos)[genos[names(ind.cond.mode),
                                                         true.qtl3] == 2]))

  betterSummary(phenos$trait1[phenos$year == last.year])
  betterSummary(phenos$trait1[phenos$year == last.year &
                              ! phenos$ind %in% resist.varieties])
  betterSummary(phenos$trait1[phenos$year == last.year &
                              phenos$ind %in% resist.varieties])

  boxplotCandidateQtl(y=ind.cond.mode,
                      X=genos[names(ind.cond.mode),],
                      snp=best.snp,
                      xlab=paste0("genotypes at ", best.snp),
                      ylab="BLUPs",
                      main="BLUPs of trait3",
                      show.points=TRUE)
}
```

### QTL detection with QTLRel

Launch QTLRel (if available):
```{r, eval=TRUE}
if(! file.exists(Sys.which("gemma"))){
  afs <- colMeans(genos) / 2
  A.mark <- genos %*% t(genos) / (2 * sum(afs * (1 - afs)))
  A.mark <- A.mark[names(ind.cond.mode), names(ind.cond.mode)]
  res3 <- list()
  res3$vc <- estVC(y=ind.cond.mode,
                   v=list(AA=A.mark, DD=NULL, HH=NULL,
                          AD=NULL, MH=NULL, EE=diag(length(ind.cond.mode))))
  idx.snps <- 1:ncol(genos)
  ## TODO: see how to specify the nb of threads
  ## TODO: scan only the chr with the causal SNP
  res3$scan <- scanOne(y=ind.cond.mode, gdat=genos[names(ind.cond.mode), idx.snps],
                       vc=res3$vc, test="F", numGeno=TRUE)
  stopifnot(all(names(res3$scan$p) == rownames(snp.coords)))
  tmp <- data.frame(BP=snp.coords$pos,
                    CHR=as.numeric(gsub("chr", "", snp.coords$chr)),
                    P=res3$scan$p,
                    SNP=rownames(snp.coords))
  print(best.snp <- as.character(tmp$SNP[which.min(tmp$P)]))
  print(tmp[tmp$SNP %in% c(true.qtl3, best.snp),])
  qqman::manhattan(x=tmp,
                   chrlabs=paste0("chr",
                                  sort(unique(as.numeric(gsub("chr", "",
                                                              snp.coords$chr))))),
                   main="QTLRel on BLUPs of trait3",
                   suggestiveline=FALSE, genomewideline=FALSE, highlight=true.qtl3)

  boxplotCandidateQtl(y=ind.cond.mode,
                      X=genos[names(ind.cond.mode),],
                      snp=best.snp,
                      xlab=paste0("genotypes at ", best.snp),
                      ylab="BLUPs",
                      main="BLUPs of trait3",
                      show.points=TRUE)
}
```


# Analysis of traits 1 and 2

## Truth

Look at the true genotypic values:
```{r}
colMeans(p0$G.A)
var(p0$G.A[,1])
var(p0$G.A[,2])
cor(p0$G.A[,1], p0$G.A[,2])
regplot(p0$G.A[,1], p0$G.A[,2], xlab="Breeding values of trait1",
        ylab="Breeding values of trait2")
```


## Separately

We want to study the "year" and "pathogen" effects on trait1 (fixed effects), and the extent to which genotypes vary in their responses (random effects, variance components).

### Trait1

#### Model fitting

Fixed-effect models:
```{r}
system.time(
    fit1f.i <- lm(trait1 ~ 1 + ind,
                  data=phenos))
system.time(
    fit1f.y.i <- lm(trait1 ~ 1 + year + ind,
                    data=phenos))
## system.time(
##     fit1f.yi <- lm(trait1 ~ 1 + year + ind + year:ind,
##                    data=phenos)) # slow
system.time(
    fit1f.y.p.i <- lm(trait1 ~ 1 + year + pathogen + ind,
                      data=phenos))
system.time(
    fit1f.y.3.i <- lm(trait1 ~ 1 + year + trait3 + ind,
                      data=phenos))
```

Mixed models (use `REML=FALSE` to compare fixed-effect parts):
```{r}
fit1m.i <- lmer(trait1 ~ 1 + (1|ind),
                data=phenos, REML=FALSE)
fit1m.y.i <- lmer(trait1 ~ 1 + year + (1|ind),
                  data=phenos, REML=FALSE)
fit1m.y.3.i <- lmer(trait1 ~ 1 + year + trait3 + (1|ind),
                    data=phenos, REML=FALSE)
```

#### Model comparison

Fixed-effect models:
```{r}
anova(fit1f.i,
      fit1f.y.i,
      fit1f.y.3.i)
AIC(fit1f.i)
AIC(fit1f.y.i)
AIC(fit1f.y.3.i)
```

Mixed models:
```{r}
anova(fit1m.i,
      fit1m.y.i,
      fit1m.y.3.i)
```

#### Model selection

```{r}
fit1 <- fit1m.y.3.i
```

#### Diagnostics

```{r}
hist(residuals(fit1) / sigma(fit1), col="grey", border="white",
     main="Scaled residuals")
abline(v=c(-2,0,2), lty=2)
plot(residuals(fit1) / sigma(fit1), fitted(fit1))
abline(v=c(-2,0,2), lty=2)
qqnorm(residuals(fit1) / sigma(fit1))
qqline(residuals(fit1) / sigma(fit1), col="red")
```

#### Inference

Parameters:
```{r}
fit1 <- lmer(formula(fit1), data=phenos, REML=TRUE)
summary(fit1)
(vc1 <- as.data.frame(VarCorr(fit1)))
vc1[vc1$grp == "ind", "vcov"] / sum(vc1[, "vcov"])
vc1[vc1$grp == "ind", "vcov"] / (vc1[vc1$grp == "ind", "vcov"] +
                                 vc1[vc1$grp == "Residual", "vcov"] /
                                 nlevels(phenos$year))
(ci1 <- confint(fit1, level=0.95, method="profile", oldNames=FALSE))
```

BLUPs of random variables:
```{r}
blups1 <- setNames(ranef(fit1)$ind[,"(Intercept)"],
                   rownames(ranef(fit1)$ind))
qqnorm(blups1, main="blups1"); qqline(blups1, col="red")
```

Compare to true values:
```{r}
cor(blups1, p0$G.A[,1], method="pearson")
cor(blups1, p0$G.A[,1], method="spearman")
regplot(p0$G.A[,1], blups1, xlab="Breeding values of trait1",
        ylab="BLUPs of genotypic values of trait1")
```

### Trait2

#### Model fitting

Mixed models (use `REML=FALSE` to compare fixed-effect parts):
```{r}
fit2m.i <- lmer(trait2 ~ 1 + (1|ind),
                data=phenos, REML=FALSE)
fit2m.y.i <- lmer(trait2 ~ 1 + year + (1|ind),
                  data=phenos, REML=FALSE)
fit2m.y.3.i <- lmer(trait2 ~ 1 + year + trait3 + (1|ind),
                    data=phenos, REML=FALSE)
```

#### Model comparison

Mixed models:
```{r}
anova(fit2m.i,
      fit2m.y.i,
      fit2m.y.3.i)
```

#### Model selection

```{r}
fit2 <- fit2m.y.i
```

#### Diagnostics

```{r}
hist(residuals(fit2) / sigma(fit2), col="grey", border="white",
     main="Scaled residuals")
abline(v=c(-2,0,2), lty=2)
plot(residuals(fit2) / sigma(fit2), fitted(fit2))
abline(v=c(-2,0,2), lty=2)
qqnorm(residuals(fit2) / sigma(fit2))
qqline(residuals(fit2) / sigma(fit2), col="red")
```

#### Inference

```{r}
fit2 <- lmer(formula(fit2), data=phenos, REML=TRUE)
summary(fit2)
(vc2 <- as.data.frame(VarCorr(fit2)))
vc2[vc2$grp == "ind", "vcov"] / sum(vc2[, "vcov"])
vc2[vc2$grp == "ind", "vcov"] / (vc2[vc2$grp == "ind", "vcov"] +
                                 vc2[vc2$grp == "Residual", "vcov"] /
                                 nlevels(phenos$year))
(ci2 <- confint(fit2, level=0.95, method="profile", oldNames=FALSE))
```

BLUPs of random variables:
```{r}
blups2 <- setNames(ranef(fit2)$ind[,"(Intercept)"],
                   rownames(ranef(fit2)$ind))
qqnorm(blups2, main="blups2"); qqline(blups2, col="red")
```

Compare to true values:
```{r}
cor(blups2, p0$G.A[,2], method="pearson")
cor(blups2, p0$G.A[,2], method="spearman")
regplot(p0$G.A[,2], blups2, xlab="Breeding values of trait2",
        ylab="BLUPs of genotypic values of trait2")
```


### Compare BLUPs for both traits

Compare BLUPs for both traits:
```{r}
all(names(blups1) == names(blups2))
cor(blups1, blups2, method="pearson")
regplot(blups1, blups2, main="Univariate LMM",
        xlab="BLUPs of genotypic values of trait1",
        ylab="BLUPs of genotypic values of trait2")
```


## Jointly

TODO: try `breedR`

Remove known confounders for both traits simultaneously (slow):
```{r, eval=FALSE}
system.time(
    fit12 <- mmer(Y=phenos[, c("trait1", "trait2")],
                  X=model.matrix(~ 1 + year + trait3, phenos),
                  Z=list(ind=list(Z=model.matrix(~ -1 + ind, phenos),
                                  K=diag(nlevels(phenos$ind)))),
                  method="NR", REML=TRUE, MVM=TRUE))
summary(fit12)
mean(fit12$u.hat$ind[,1])
mean(fit12$u.hat$ind[,2])
var(fit12$u.hat$ind[,1])
var(fit12$u.hat$ind[,2])
cor(fit12$u.hat$ind[,1], fit12$u.hat$ind[,2])
regplot(fit12$u.hat$ind[,1], fit12$u.hat$ind[,2], main="Multivariate LMM",
        xlab="Expected breeding values of trait1",
        ylab="Expected breeding values of trait2")
```


# Genomic prediction

## Generic functions

The `cvTools` package requires the called function to return a `S3` class with a `predict` method.
So here it is for the ridge regression as implemented in the `rrBLUP` package:
```{r}
rr <- function(y, Z, K=NULL, X=NULL, method="REML"){
  out <- mixed.solve(y=y, Z=Z, K=K, X=X, method=method)
  return(structure(out, class="rr"))
}
predict.rr <- function(object, newZ){
  out <- as.vector(newZ %*% object$u)
  if(! is.null(rownames(newZ)))
    names(out) <- rownames(newZ)
  return(out)
}
```

Set up the folds:
```{r}
set.seed(1859)
folds <- cvFolds(n=nrow(genos), K=5, R=10)
dim(folds$subsets)
head(folds$subsets)
```

## Trait1

### Ridge regression on the full data set

Fit the ridge regression model:
```{r, eval=TRUE}
fit <- mixed.solve(y=blups1[rownames(genos)], Z=genos)
fit$Vu
fit$Ve
## TODO: calc PVE
```

Compare estimated SNP effects with their true values:
```{r, eval=TRUE}
cor(p0$Beta[colnames(genos), "trait1"], fit$u)
regplot(x=p0$Beta[colnames(genos), "trait1"], y=fit$u,
        xlab="true SNP effects", ylab="estimated SNP effects",
        main="Ridge regression for BLUPs of trait1")
```

Compare GEBVs (predicted BLUPs of breeding values) with EBVs (true BLUPs of genotypic values) and with BVs (breeding values, i.e. additive genotypic values):
```{r, eval=TRUE}
cor(blups1[rownames(genos)], genos %*% fit$u)
regplot(x=blups1[rownames(genos)], y=genos %*% fit$u,
        xlab="BLUPs of genotypic values", ylab="predicted GBLUPs",
        main="Ridge regression for BLUPs of trait1")
cor(p0$G.A[rownames(genos), "trait1"], genos %*% fit$u)
regplot(x=p0$G.A[rownames(genos), "trait1"], y=genos %*% fit$u,
        xlab="true genotypic values", ylab="predicted GBLUPs",
        main="Ridge regression for BLUPs of trait1")
```

### Cross-validation

Set up the function call:
```{r, eval=TRUE}
call <- call("rr", y=blups1[rownames(genos)], Z=genos)
```

Perform cross-validation:
```{r trait1_crossval, eval=TRUE}
system.time(
    fit.cv <- cvTool(call=call, x=genos, y=blups1[rownames(genos)],
                     names=c("Z", "y"), cost=rmspe, folds=folds))
fit.cv
summary(fit.cv)
sd(fit.cv[,1])
fit.cv <- cvTool(call=call, x=genos, y=blups1[rownames(genos)],
                 names=c("Z", "y"), cost=cor, folds=folds)
fit.cv
summary(fit.cv)
sd(fit.cv[,1])
```


## Trait2

### Ridge regression on the full data set

Fit the ridge regression model:
```{r, eval=TRUE}
fit <- mixed.solve(y=blups2[rownames(genos)], Z=genos)
fit$Vu
fit$Ve
## TODO: calc PVE
```

Compare estimated SNP effects with their true values:
```{r, eval=TRUE}
cor(p0$Beta[colnames(genos), "trait2"], fit$u)
regplot(x=p0$Beta[colnames(genos), "trait2"], y=fit$u,
        xlab="true SNP effects", ylab="estimated SNP effects",
        main="Ridge regression for BLUPs of trait2")
```

Compare GEBVs (predicted BLUPs of breeding values) with EBVs (true BLUPs of genotypic values) and with BVs (breeding values, i.e. additive genotypic values):
```{r, eval=TRUE}
cor(blups2[rownames(genos)], genos %*% fit$u)
regplot(x=blups2[rownames(genos)], y=genos %*% fit$u,
        xlab="BLUPs of genotypic values", ylab="predicted GBLUPs",
        main="Ridge regression for BLUPs of trait2")
regplot(x=p0$G.A[rownames(genos), "trait2"], y=genos %*% fit$u,
        xlab="true genotypic values", ylab="predicted GBLUPs",
        main="Ridge regression for BLUPs of trait2")
```

### Cross-validation

Set up the function call:
```{r, eval=TRUE}
call <- call("rr", y=blups2[rownames(genos)], Z=genos)
```

Perform cross-validation:
```{r trait2_crossval, eval=TRUE}
system.time(
    fit.cv <- cvTool(call=call, x=genos, y=blups2[rownames(genos)],
                     names=c("Z", "y"), cost=rmspe, folds=folds))
fit.cv
summary(fit.cv)
fit.cv <- cvTool(call=call, x=genos, y=blups2[rownames(genos)],
                 names=c("Z", "y"), cost=cor, folds=folds)
fit.cv
summary(fit.cv)
sd(fit.cv[,2])
```


# Appendix

```{r info}
t1 <- proc.time(); t1 - t0
print(sessionInfo(), locale=FALSE)
```
